libraries/core_libs/consensus/src/pillar_chain/pillar_chain_manager.cpp

```diff
@@ -0,0 +1,475 @@
+    last_finalized_pillar_block_ = std::move(latest_pillar_block_data->block_);
+  }
+
+  LOG_OBJECTS_CREATE("PILLAR_CHAIN");
+}
+
+std::shared_ptr<PillarBlock> PillarChainManager::createPillarBlock(
+    const std::shared_ptr<final_chain::FinalizationResult>& block_data) {
+  const auto block_num = block_data->final_chain_blk->number;
+
+  PillarBlock::Hash previous_pillar_block_hash{};  // null block hash
+  auto new_stakes = final_chain_->dpos_validators_total_stakes(block_num);
+  std::vector<PillarBlock::ValidatorStakeChange> stakes_changes;
+
+  // First ever pillar block
+  if (block_num == kFicusHfConfig.firstPillarBlockPeriod()) {
+    // First pillar block - use all current stakes as changes
+    stakes_changes.reserve(new_stakes.size());
+    std::transform(new_stakes.begin(), new_stakes.end(), std::back_inserter(stakes_changes),
+                   [](auto& stake) { return PillarBlock::ValidatorStakeChange(stake); });
+  } else {
+    // Second or further pillar blocks
+    if (!current_pillar_block_) {
+      LOG(log_er_) << "Empty previous pillar block, new pillar block period " << block_num;
+      assert(false);
+      return nullptr;
+    }
+
+    // Push pillar block into the pillar chain in case it was not finalized yet
+    if (!isPillarBlockLatestFinalized(current_pillar_block_->getHash())) {
+      // Get 2t+1 verified votes
+      auto two_t_plus_one_votes =
+          pillar_votes_.getVerifiedVotes(current_pillar_block_->getPeriod(), current_pillar_block_->getHash(), true);
+      if (two_t_plus_one_votes.empty()) {
+        LOG(log_er_) << "There is < 2t+1 votes for current pillar block " << current_pillar_block_->getHash()
+                     << ", period: " << current_pillar_block_->getPeriod() << ". Current period " << block_num;
+        return nullptr;
+      }
+
+      // Save current pillar block and 2t+1 votes into db
+      if (!pushPillarBlock(PillarBlockData{current_pillar_block_, std::move(two_t_plus_one_votes)})) {
+        // This should never happen
+        LOG(log_er_) << "Unable to push pillar block: " << current_pillar_block_->getHash() << ", period "
+                     << current_pillar_block_->getPeriod();
+        assert(false);
+        return nullptr;
+      }
+    }
+
+    // This should never happen
+    if (current_pillar_block_stakes_.empty()) {
+      LOG(log_er_) << "Empty current pillar block stakes, new pillar block period " << block_num;
+      assert(false);
+      return nullptr;
+    }
+
+    previous_pillar_block_hash = current_pillar_block_->getHash();
+
+    // Get validators stakes changes between the current and previous pillar block
+    stakes_changes = getOrderedValidatorsStakesChanges(new_stakes, current_pillar_block_stakes_);
+  }
+
+  // TODO: provide bridge root ???
+  const auto pillar_block = std::make_shared<PillarBlock>(block_num, block_data->final_chain_blk->state_root, h256{},
```

Comment:
To discuss: Replace bridge root with structure, so it will be easier to extend it later

Reply:
this is being worked on here I guess ? https://github.com/Taraxa-project/taraxa-node/pull/2746